Improvement/cldsrv 427 permissions check

benzekrimaha opened the pull request:
PR opened after closing : https://github.com/scality/cloudserver/pull/5323

Bucket policies are not correctly interpreted, this is part of the following epic to fix that: https://github.com/scality/Arsenal/pull/2181

This PR is aiming enable authorization logics from IAM policies, bucket policies and ACLs to be interpreted in aggregate in CloudServer's 'isBucketAuthorized' and 'isObjectAuthorized' functions.  , ticket linked to this issue here : https://scality.atlassian.net/browse/CLDSRV-427

PRs providing implicit Deny logic to CS for processing in this PR
https://github.com/scality/Arsenal/pull/2181
https://github.com/scality/Vault/pull/2135
https://github.com/scality/cloudserver/pull/5322
https://github.com/scality/cloudserver/pull/5420

I'm not bumping a new CLDSRV version since a new version has been created in this merged PR : https://github.com/scality/cloudserver/pull/5322 , Please let me know if it needs to be done anyways.

Commit:
CLDSRV-426: update ACL permission checks for implicitDeny logic

CLDSRV-426:fixups on ACL permission checks for implicitDeny logic

CLDSRV-426:better readability on ACL permission

Commit:
CLDSRV-426: add tests for ACL permission check updates

CLDSRV-426: additionnal test for ACL permission

Commit:
CLDSRV-427: update bucket/object perm checks to account for implicit …
…denies

(cherry picked from commit c01898f1a002508e9c29662a2a5b07a4b0a26dc8)

Commit:
CLDSRV-424: api call updated with implicit deny logic

change variable names for clarity

edit: update arsenal package

benzekrimaha committed:
CLDSRV-427: linting fixups and retrocompatibility changes

benzekrimaha committed:
CLDSRV-424:ARSN version bump

benzekrimaha committed:
CLDSRV-424:CLDSRV version bump

Update lib/api/api.js

Co-authored-by: Jonathan Gramain <jonathan.gramain@scality.com>

benzekrimaha committed:
CLDSRV-427: functions refactor

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-427 contains:


* `7.70.30`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.33`

* `7.70.30`

* `8.6.12`

* `8.7.31`

* `8.8.5`


Please check the `Fix Version/s` of CLDSRV-427, or the target
branch of this pull request.

KazToozs commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
-        return false;
-    }
-    return (aclPermission || (bucketPolicyPermission === 'allow'));
+        if (method === 'isBucketAuthorized') {
```
I'm of the personal opinion that if such conditions are necessary for such large parts of code, it's better practice to keep this separate into smaller, different functions.

benzekrimaha replied:
Actually this check was only added for the bucket specific checks/ object ones in the else , as the functions are nearly identical besides for those checks I used the method param passed by the functions 'isBucketAUthorized', 'isObjectAuthorized' and 'evaluateBucketPolicyWithIAM' to separate them and keep all of the changes related to authorizations in only one function (can ease modifications in the future as well)

KazToozs reviewed (commented):
Code makes sense otherwise

welansari commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
+        // eslint-disable-next-line no-param-reassign
+        actionImplicitDenies = {};
+    }
+    requestTypes.forEach(requestType => {
+        if (actionImplicitDenies[requestType] === undefined) {
+            // eslint-disable-next-line no-param-reassign
+            actionImplicitDenies[requestType] = false;
+        }
+    });
```
No need to initialize `undefined` values of `actionImplicitDenies`.
Inside the second `forEach` you can directly test if it's false with `!!actionImplicitDenies[requestType]`.

welansari commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
-            return true;
+    const results = {};
+    const mainApiCall = requestTypes[0];
+    requestTypes.forEach(_requestType => {
```
You can directly return `requestTypes.every(_requestType => {...})` instead of storing and checking all the results.
This will also avoid looping over everything when you encounter a first error.

Kerkesni commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+            // account is authorized if:
+            // - requesttype is included in bucketOwnerActions and
+            // - account is the bucket owner
+            // - requester is account, not user
+            if (bucketOwnerActions.includes(parsedMethodName)
+            && (bucketOwner === canonicalID)
+            && requesterIsNotUser) {
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+        }
+        const aclPermission = method === 'isBucketAuthorized' ?
+            checkBucketAcls(bucket, _requestType, canonicalID, mainApiCall) :
+            checkObjectAcls(bucket, objectMD, parsedMethodName, canonicalID,
+                requesterIsNotUser, isUserUnauthenticated, mainApiCall);
+        const bucketPolicy = bucket.getBucketPolicy();
+        if (!bucketPolicy) {
+            results[_requestType] = withIam === true ? actionImplicitDenies[_requestType] === false :
+            actionImplicitDenies[_requestType] === false && aclPermission;
+            return;
+        }
+        const bucketPolicyPermission = checkBucketPolicy(bucketPolicy, _requestType, canonicalID,
+            arn, bucketOwner, log, request);
+
+        if (bucketPolicyPermission === 'explicitDeny') {
+            results[_requestType] = false;
+            return;
+        }
+        // If the bucket policy returns an allow, we accept the request, as the
+        // IAM response here is either Allow or implicit deny.
+        if (bucketPolicyPermission === 'allow') {
+            results[_requestType] = true;
+            return;
+        }
+        // eslint-disable-next-line max-len
+        results[_requestType] = withIam === true ? actionImplicitDenies[_requestType] === false :
+            actionImplicitDenies[_requestType] === false && aclPermission;
```
I find it hard to understand what's going on compared to the old code, i suggest using more helper functions that separate the logic and that don't depend on the caller method

welansari commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
-    return (aclPermission || (bucketPolicyPermission === 'allow'));
-}
+function isAuthorized(method, bucket, objectMD, requestTypes, canonicalID, authInfo, log, request,
+    actionImplicitDenies, withIam) {
+    if (!withIam) {
+        // eslint-disable-next-line no-param-reassign
+        withIam = false;
+    }
```
```suggestion
    actionImplicitDenies, withIam = false) {
```

benzekrimaha replied:
As I'm keeping 3 separate functions , this param will no longer be needed

welansari commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
+    }
+    // By default, all missing actions are defined as allowed from IAM, to be
+    // backward compatible
+    if (!actionImplicitDenies) {
+        // eslint-disable-next-line no-param-reassign
+        actionImplicitDenies = {};
+    }
```
Same as for `withIam`, we can have a default value equal to `{}` in the function parameter declaration

benzekrimaha replied:
This checks also when we call the function with false as a parameter for actionImplicitDenies so putting the value by default will prevent assigning the empty object when false is passed to the function

welansari commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -268,75 +268,120 @@ function checkBucketPolicy(policy, requestType, canonicalID, arn, bucketOwner, l
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+            // account is authorized if:
+            // - requesttype is included in bucketOwnerActions and
+            // - account is the bucket owner
+            // - requester is account, not user
+            if (bucketOwnerActions.includes(parsedMethodName)
+            && (bucketOwner === canonicalID)
+            && requesterIsNotUser) {
+                results[_requestType] = actionImplicitDenies[_requestType] === false;
+                return;
+            }
+        }
+        const aclPermission = method === 'isBucketAuthorized' ?
+            checkBucketAcls(bucket, _requestType, canonicalID, mainApiCall) :
+            checkObjectAcls(bucket, objectMD, parsedMethodName, canonicalID,
+                requesterIsNotUser, isUserUnauthenticated, mainApiCall);
+        const bucketPolicy = bucket.getBucketPolicy();
+        if (!bucketPolicy) {
+            results[_requestType] = withIam === true ? actionImplicitDenies[_requestType] === false :
+            actionImplicitDenies[_requestType] === false && aclPermission;
+            return;
+        }
+        const bucketPolicyPermission = checkBucketPolicy(bucketPolicy, _requestType, canonicalID,
+            arn, bucketOwner, log, request);
+
+        if (bucketPolicyPermission === 'explicitDeny') {
+            results[_requestType] = false;
+            return;
+        }
+        // If the bucket policy returns an allow, we accept the request, as the
+        // IAM response here is either Allow or implicit deny.
+        if (bucketPolicyPermission === 'allow') {
+            results[_requestType] = true;
+            return;
+        }
+        // eslint-disable-next-line max-len
+        results[_requestType] = withIam === true ? actionImplicitDenies[_requestType] === false :
+            actionImplicitDenies[_requestType] === false && aclPermission;
```
I find it hard to understand what's going on compared to the old code, i suggest using more helper functions that separate the logic and that don't depend on the caller method

benzekrimaha replied:
I guess I'll separate the functions (as before)  and just do fixups instead of refactoring the 3 in one function as it will be easier to read/understand

benzekrimaha committed:
CLDSRV-427: Improving functions using helper function

- In this commit , I added a helper (processBucketPolicy) function
for the bycket policies checks that are shared between the
isbucketAuthorized, isObjAuthorized and evaluateBucketPolicyWithIAM
for a better code readability and to avoid long functions.
